watchers/deployment_handler.go

```diff
@@ -0,0 +1,231 @@
+func (h DeploymentHandler) OnUpdate(oldObj, newObj interface{}) {
+	// Convert oldObj and newObj to Deployment objects
+	oldDeployment, okOldDeployment := oldObj.(*appsv1.Deployment)
+	newDeployment, okNewDeployment := newObj.(*appsv1.Deployment)
+
+	// If both old and new are not deployments, do nothing
+	if !okOldDeployment || !okNewDeployment {
+		return
+	}
+
+	// If deployment doesn't have associated disruption rollout, do nothing
+	hasDisruptionRollout, err := h.HasAssociatedDisruptionRollout(newDeployment)
+	if !hasDisruptionRollout || err != nil {
+		return
+	}
+
+	// If containers have't changed, do nothing
+	containersChanged, initContainersHash, containersHash, err := ContainersChanged(&oldDeployment.Spec.Template.Spec, &newDeployment.Spec.Template.Spec, h.log)
+	if !containersChanged || err != nil {
+		return
+	}
+
+	err = h.UpdateDisruptionRolloutStatus(newDeployment, initContainersHash, containersHash)
+	if err != nil {
+		return
+	}
+}
+
+// OnDelete is a handler function for the delete of a deployment
+func (h DeploymentHandler) OnDelete(_ interface{}) {
+	// Do nothing on delete event
+}
+
+func (h DeploymentHandler) FetchAssociatedDisruptionRollouts(deployment *appsv1.Deployment) (*chaosv1beta1.DisruptionRolloutList, error) {
+	indexedValue := "deployment" + "-" + deployment.Namespace + "-" + deployment.Name
+
+	disruptionRollouts := &chaosv1beta1.DisruptionRolloutList{}
+	err := h.Client.List(context.Background(), disruptionRollouts, client.MatchingFields{"targetResource": indexedValue})
+
+	if err != nil {
+		h.log.Errorw("unable to fetch DisruptionRollouts using index", "error", err, "indexedValue", indexedValue)
+		return nil, err
+	}
+
+	return disruptionRollouts, nil
+}
+
+func (h DeploymentHandler) HasAssociatedDisruptionRollout(deployment *appsv1.Deployment) (bool, error) {
+	disruptionRollouts, err := h.FetchAssociatedDisruptionRollouts(deployment)
+	if err != nil {
+		h.log.Errorw("unable to check for associated DisruptionRollout", "Deployment", deployment.Name, "error", err)
+		return false, err
+	}
+
+	return len(disruptionRollouts.Items) > 0, nil
+}
+
+func (h DeploymentHandler) UpdateDisruptionRolloutStatus(deployment *appsv1.Deployment, initContainersHash, containersHash map[string]string) error {
+	disruptionRollouts, err := h.FetchAssociatedDisruptionRollouts(deployment)
+	if err != nil {
+		return err
+	}
+
+	for _, dr := range disruptionRollouts.Items {
+		dr.Status.LatestInitContainersHash = initContainersHash
+		dr.Status.LatestContainersHash = containersHash
+		dr.Status.LastContainerChangeTime = &metav1.Time{Time: time.Now()}
+
+		err = h.Client.Status().Update(context.TODO(), &dr)
+		if err != nil {
+			h.log.Errorw("unable to update DisruptionRollout status", "DisruptionRollout", dr.Name, "error", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func ContainersChanged(oldPodSpec, newPodSpec *corev1.PodSpec, log *zap.SugaredLogger) (bool, map[string]string, map[string]string, error) {
```

Comment:
Let's move these functions used by both deployment_handler and statefulset_handler, out of deployment_handler.go